File: oriedita/src/main/java/com/example/oriedita/editor/handler/MouseHandlerBackgroundChangePosition.java
package com.example.oriedita.editor.handler;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import oriedita.editor.canvas.MouseMode;
import oriedita.editor.databinding.BackgroundModel;
import oriedita.editor.service.ResetService;
import origami.crease_pattern.elements.Rectangle;
import origami.crease_pattern.elements.LineColor;
import origami.crease_pattern.elements.LineSegment;
import origami.crease_pattern.elements.Point;


@ApplicationScoped
@Handles(MouseMode.BACKGROUND_CHANGE_POSITION_26)
public class MouseHandlerBackgroundChangePosition extends BaseMouseHandler {
    private final ResetService resetService;
    private final BackgroundModel backgroundModel;

    @Override
    public void mouseMoved(Point p0) {

    }

    //マウス操作(ボタンを押したとき)時の作業
    public void mousePressed(Point p0) {
        Point p = d.getCamera().TV2object(p0);

        if (d.getLineStep().size() == 3) {
            Point closestPoint = d.getClosestPoint(p);

            if (p.distance(closestPoint) < d.getSelectionDistance()) {
                p = closestPoint;
            }
            d.lineStepAdd(new LineSegment(p, p, LineColor.ORANGE_4));
        } else if (d.getLineStep().size() == 2) {
            Point closestPoint = d.getClosestPoint(p);

            if (p.distance(closestPoint) < d.getSelectionDistance()) {
                p = closestPoint;
            }
            d.lineStepAdd(new LineSegment(p, p, LineColor.CYAN_3));
        } else if (d.getLineStep().size() == 1) {
            d.lineStepAdd(new LineSegment(p, p, LineColor.BLUE_2));
        } else if (d.getLineStep().size() == 0) {
            d.lineStepAdd(new LineSegment(p, p, LineColor.RED_1));
        }
    }

    //マウス操作(ドラッグしたとき)を行う関数
    public void mouseDragged(Point p0) {
    }

    @Inject
    public MouseHandlerBackgroundChangePosition(ResetService resetService, BackgroundModel backgroundModel) {
        this.resetService = resetService;
        this.backgroundModel = backgroundModel;
    }

    //マウス操作(ボタンを離したとき)を行う関数
    public void mouseReleased(Point p0) {
        if (d.getLineStep().size() == 4) {
            LineSegment s_1 = d.getLineStep().get(0);
            LineSegment s_2 = d.getLineStep().get(1);
            LineSegment s_3 = d.getLineStep().get(2);
            LineSegment s_4 = d.getLineStep().get(3);
            resetService.Button_shared_operation();

            backgroundModel.setLockBackground(false);

            backgroundModel.setBackgroundPosition(new Rectangle(d.getCamera().object2TV(s_1.getA()),
                    d.getCamera().object2TV(s_2.getA()),
                    d.getCamera().object2TV(s_3.getA()),
                    d.getCamera().object2TV(s_4.getA())));
        }
    }
}

File: oriedita-ui/src/main/java/com/example/oriedita_ui/handler/MouseHandlerBackgroundChangePosition.kt
package com.example.oriedita_ui.handler

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerInputChange
import com.example.oriedita_core.origami.crease_pattern.elements.LineSegment
import com.example.oriedita_core.origami.crease_pattern.elements.Point
import com.example.oriedita_core.origami.crease_pattern.elements.LineColor
import com.example.oriedita_core.origami.crease_pattern.elements.Rectangle

/**
 * Обработчик изменения позиции фона
 * Адаптированная версия для Android
 */
class MouseHandlerBackgroundChangePosition : BaseMouseHandler() {
    
    private var stepPoints = mutableListOf<Point>()
    private var stepSegments = mutableListOf<LineSegment>()
    private var isProcessing = false
    private var backgroundRectangle: Rectangle? = null
    
    override fun onPress(offset: Offset, change: PointerInputChange): Boolean {
        val point = offsetToPoint(offset)
        
        // Добавить точку в зависимости от количества уже добавленных точек
        when (stepPoints.size) {
            0 -> {
                addStepPoint(point, LineColor.RED_1)
                println("Добавлена первая точка фона: (${point.x}, ${point.y})")
            }
            1 -> {
                addStepPoint(point, LineColor.BLUE_2)
                println("Добавлена вторая точка фона: (${point.x}, ${point.y})")
            }
            2 -> {
                addStepPoint(point, LineColor.CYAN_3)
                println("Добавлена третья точка фона: (${point.x}, ${point.y})")
            }
            3 -> {
                addStepPoint(point, LineColor.ORANGE_4)
                println("Добавлена четвертая точка фона: (${point.x}, ${point.y})")
            }
        }
        
        return true
    }
    
    override fun onDrag(offset: Offset, change: PointerInputChange): Boolean {
        // Изменение позиции фона не требует перетаскивания
        return false
    }
    
    override fun onRelease(offset: Offset, change: PointerInputChange): Boolean {
        if (stepPoints.size == 4) {
            // Создать прямоугольник фона из четырех точек
            createBackgroundRectangle()
            reset()
            println("Позиция фона изменена")
        }
        
        return true
    }
    
    private fun addStepPoint(point: Point, color: LineColor) {
        val closestPoint = getClosestPoint(point)
        val finalPoint = if (point.distance(closestPoint) < getSelectionDistance()) {
            closestPoint
        } else {
            point
        }
        
        stepPoints.add(finalPoint)
        val segment = LineSegment(finalPoint, finalPoint, color)
        stepSegments.add(segment)
        
        // Добавить сегмент в линию шага
        addLineStep(segment)
    }
    
    private fun createBackgroundRectangle() {
        if (stepPoints.size != 4) return
        
        val p1 = stepPoints[0]
        val p2 = stepPoints[1]
        val p3 = stepPoints[2]
        val p4 = stepPoints[3]
        
        // Создать прямоугольник из четырех точек
        backgroundRectangle = Rectangle(p1, p2, p3, p4)
        
        // Установить позицию фона
        setBackgroundPosition(backgroundRectangle!!)
        
        // Разблокировать фон
        setBackgroundLock(false)
        
        println("Создан прямоугольник фона: ${getRectangleDescription(backgroundRectangle!!)}")
    }
    
    private fun setBackgroundPosition(rectangle: Rectangle) {
        // Здесь должна быть логика установки позиции фона
        println("Установлена позиция фона: ${getRectangleDescription(rectangle)}")
    }
    
    private fun setBackgroundLock(locked: Boolean) {
        // Здесь должна быть логика блокировки/разблокировки фона
        println("Фон ${if (locked) "заблокирован" else "разблокирован"}")
    }
    
    private fun getRectangleDescription(rectangle: Rectangle): String {
        val p1 = rectangle.getP1()
        val p2 = rectangle.getP2()
        val p3 = rectangle.getP3()
        val p4 = rectangle.getP4()
        
        return "Прямоугольник: P1(${String.format("%.2f", p1.x)}, ${String.format("%.2f", p1.y)}), " +
               "P2(${String.format("%.2f", p2.x)}, ${String.format("%.2f", p2.y)}), " +
               "P3(${String.format("%.2f", p3.x)}, ${String.format("%.2f", p3.y)}), " +
               "P4(${String.format("%.2f", p4.x)}, ${String.format("%.2f", p4.y)})"
    }
    
     override fun reset() {
        stepPoints.clear()
        stepSegments.clear()
        isProcessing = false
        clearLineStep()
        
        println("Обработчик изменения позиции фона сброшен")
    }
    
    fun getStepPoints(): List<Point> = stepPoints.toList()
    
    fun getStepSegments(): List<LineSegment> = stepSegments.toList()
    
    fun getStepCount(): Int = stepPoints.size
    
    fun getBackgroundRectangle(): Rectangle? = backgroundRectangle
    
    fun isProcessing(): Boolean = isProcessing
    
    fun getBackgroundPositionDescription(): String {
        return when (stepPoints.size) {
            0 -> "Выберите первую точку фона"
            1 -> "Выберите вторую точку фона"
            2 -> "Выберите третью точку фона"
            3 -> "Выберите четвертую точку фона"
            4 -> "Позиция фона установлена"
            else -> "Неизвестное состояние"
        }
    }
    
    fun getBackgroundRectangleDescription(): String {
        return backgroundRectangle?.let { getRectangleDescription(it) } ?: "Прямоугольник фона не создан"
    }
    
    override fun getName(): String = "Изменение позиции фона"
    
    override fun getDescription(): String = "Установите новую позицию фона, выбрав четыре точки"
} 